facebook/react-native
# Add ability to change cache policy on iOS

React Native does not currently expose the ability to decide on the cache policy used by network requests. Android can work around this by utilising OkHttp Interceptors. For iOS, this PR extends the functionality of `RCTHTTPRequestHandler` by exposing an JS interface that can alter the cache policy as defined in [`NSURLRequestCachePolicy`](https://developer.apple.com/library/ios/documentation/Cocoa/Reference/Foundation/Classes/NSURLRequest_Class/#//apple_ref/c/tdef/NSURLRequestCachePolicy)

**Test plan**

- Make a request that returns a response with a relatively short cache duration
- Set the cache policy to `ReturnCacheDataDontLoad`
- Go offline, wait until the cached response should have expired, then make the request again.

The XMLHttpRequest demo in the UIExplorer example project has been updated with an example that demonstrates the functionality.

**Example Usage**

If a client wants to be able to serve previously viewed content when offline, even if it is known it could be stale, they could achieve it in such a way:

```
const { NetInfo } = require('react-native');
const { setCachePolicy } = require('NativeModules').HTTPRequestHandler;

NetInfo.isConnected.addEventListener(
  'change',
  (isConnected) => {
    if (isConnected) {
      setCachePolicy('UseProtocolCachePolicy')
    } else {
      setCachePolicy('ReturnCacheDataDontLoad')
    }
  }
);

```
